File: WellnessWingman/Services/Navigation/NavigationBreadcrumb.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace HealthHelper.Services.Navigation;

public sealed class NavigationBreadcrumb
{
    private static readonly Dictionary<HistoricalViewLevel, string> RouteMap = new()
    {
        { HistoricalViewLevel.Today, "today" },
        { HistoricalViewLevel.Week, "week" },
        { HistoricalViewLevel.Month, "month" },
        { HistoricalViewLevel.Year, "year" },
        { HistoricalViewLevel.Day, "day" }
    };

    public NavigationBreadcrumb(HistoricalViewLevel level, DateTime date, string? label = null)
    {
        Level = level;
        Date = date;
        Label = label ?? BuildDefaultLabel(level, date);
    }

    public HistoricalViewLevel Level { get; }

    public DateTime Date { get; }

    public string Label { get; }

    public string Route => RouteMap[Level];

    public IDictionary<string, object> BuildNavigationParameters()
    {
        var parameters = new Dictionary<string, object>();
        switch (Level)
        {
            case HistoricalViewLevel.Day:
                parameters["Date"] = Date;
                break;
            case HistoricalViewLevel.Week:
                parameters["WeekStart"] = Date;
                break;
            case HistoricalViewLevel.Month:
                parameters["Year"] = Date.Year;
                parameters["Month"] = Date.Month;
                break;
            case HistoricalViewLevel.Year:
                parameters["Year"] = Date.Year;
                break;
        }

        return parameters;
    }

    private static string BuildDefaultLabel(HistoricalViewLevel level, DateTime date)
    {
        // TODO: Localise breadcrumb labels once shared resources are available.
        return level switch
        {
            HistoricalViewLevel.Today => "Today",
            HistoricalViewLevel.Day => date.ToString("D", CultureInfo.CurrentCulture),
            HistoricalViewLevel.Week => $"Week of {date:MMM d}",
            HistoricalViewLevel.Month => date.ToString("MMMM yyyy", CultureInfo.CurrentCulture),
            HistoricalViewLevel.Year => date.ToString("yyyy", CultureInfo.InvariantCulture),
            _ => date.ToString(CultureInfo.CurrentCulture)
        };
    }
}

File: kotlin/shared/src/commonMain/kotlin/com/wellnesswingman/domain/navigation/NavigationBreadcrumb.kt
package com.wellnesswingman.domain.navigation

import kotlinx.datetime.LocalDate
import kotlinx.datetime.Month

/**
 * Represents a point in the navigation history that can be returned to.
 */
data class NavigationBreadcrumb(
    val level: HistoricalViewLevel,
    val date: LocalDate,
    val label: String = buildDefaultLabel(level, date)
) {
    /**
     * Route name for this breadcrumb.
     */
    val route: String = level.route

    /**
     * Builds navigation parameters for this breadcrumb.
     */
    fun buildNavigationParameters(): Map<String, Any> {
        val parameters = mutableMapOf<String, Any>()
        when (level) {
            HistoricalViewLevel.DAY -> {
                parameters["date"] = date.toString()
            }
            HistoricalViewLevel.WEEK -> {
                parameters["weekStart"] = date.toString()
            }
            HistoricalViewLevel.MONTH -> {
                parameters["year"] = date.year
                parameters["month"] = date.monthNumber
            }
            HistoricalViewLevel.YEAR -> {
                parameters["year"] = date.year
            }
            HistoricalViewLevel.TODAY -> {
                // No parameters needed
            }
        }
        return parameters
    }

    companion object {
        private fun buildDefaultLabel(level: HistoricalViewLevel, date: LocalDate): String {
            return when (level) {
                HistoricalViewLevel.TODAY -> "Today"
                HistoricalViewLevel.DAY -> formatDayLabel(date)
                HistoricalViewLevel.WEEK -> "Week of ${formatMonthDay(date)}"
                HistoricalViewLevel.MONTH -> "${date.month.name.lowercase().replaceFirstChar { it.uppercase() }} ${date.year}"
                HistoricalViewLevel.YEAR -> "${date.year}"
            }
        }

        private fun formatDayLabel(date: LocalDate): String {
            val monthName = date.month.name.lowercase().replaceFirstChar { it.uppercase() }
            return "$monthName ${date.dayOfMonth}, ${date.year}"
        }

        private fun formatMonthDay(date: LocalDate): String {
            val monthAbbrev = date.month.name.take(3).lowercase().replaceFirstChar { it.uppercase() }
            return "$monthAbbrev ${date.dayOfMonth}"
        }
    }
}
